### ISSUE TITLE
Flaky-test: PulsarFunctionLocalRunTest

### ISSUE THREAD

**Issue by lhotari** [apache/pulsar#8720]
PulsarFunctionLocalRunTest is flaky. The setup method fails sporadically.

[example failure](https://github.com/apache/pulsar/pull/8707/checks?check_run_id=1457593039#step:9:171)

```
Error:  Tests run: 14, Failures: 3, Errors: 0, Skipped: 3, Time elapsed: 169.285 s <<< FAILURE! - in org.apache.pulsar.functions.worker.PulsarFunctionLocalRunTest
Error:  setup(org.apache.pulsar.functions.worker.PulsarFunctionLocalRunTest)  Time elapsed: 63.889 s  <<< FAILURE!
org.apache.pulsar.broker.PulsarServerException: java.lang.RuntimeException: org.apache.pulsar.client.admin.PulsarAdminException: java.lang.NullPointerException
	at org.apache.pulsar.broker.PulsarService.start(PulsarService.java:611)
	at org.apache.pulsar.functions.worker.PulsarFunctionLocalRunTest.setup(PulsarFunctionLocalRunTest.java:184)
```

<details>
<summary>Full exception stacktrace</summary>
<code><pre>
Error:  Tests run: 14, Failures: 3, Errors: 0, Skipped: 3, Time elapsed: 169.285 s <<< FAILURE! - in org.apache.pulsar.functions.worker.PulsarFunctionLocalRunTest
Error:  setup(org.apache.pulsar.functions.worker.PulsarFunctionLocalRunTest)  Time elapsed: 63.889 s  <<< FAILURE!
org.apache.pulsar.broker.PulsarServerException: java.lang.RuntimeException: org.apache.pulsar.client.admin.PulsarAdminException: java.lang.NullPointerException
	at org.apache.pulsar.broker.PulsarService.start(PulsarService.java:611)
	at org.apache.pulsar.functions.worker.PulsarFunctionLocalRunTest.setup(PulsarFunctionLocalRunTest.java:184)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.MethodInvocationHelper.invokeMethodConsideringTimeout(MethodInvocationHelper.java:61)
	at org.testng.internal.ConfigInvoker.invokeConfigurationMethod(ConfigInvoker.java:366)
	at org.testng.internal.ConfigInvoker.invokeConfigurations(ConfigInvoker.java:320)
	at org.testng.internal.TestInvoker.runConfigMethods(TestInvoker.java:701)
	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:527)
	at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:174)
	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
	at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:822)
	at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:147)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.testng.TestRunner.privateRun(TestRunner.java:764)
	at org.testng.TestRunner.run(TestRunner.java:585)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:384)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:378)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:337)
	at org.testng.SuiteRunner.run(SuiteRunner.java:286)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:53)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:96)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1218)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1140)
	at org.testng.TestNG.runSuites(TestNG.java:1069)
	at org.testng.TestNG.run(TestNG.java:1037)
	at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:135)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.executeSingleClass(TestNGDirectoryTestSuite.java:112)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.executeLazy(TestNGDirectoryTestSuite.java:123)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.execute(TestNGDirectoryTestSuite.java:90)
	at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:146)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
Caused by: java.lang.RuntimeException: org.apache.pulsar.client.admin.PulsarAdminException: java.lang.NullPointerException
	at org.apache.pulsar.functions.worker.WorkerService.start(WorkerService.java:296)
	at org.apache.pulsar.broker.PulsarService.startWorkerService(PulsarService.java:1311)
	at org.apache.pulsar.broker.PulsarService.start(PulsarService.java:596)
	... 39 more
Caused by: org.apache.pulsar.client.admin.PulsarAdminException: java.lang.NullPointerException
	at org.apache.pulsar.client.admin.internal.BaseResource.getApiException(BaseResource.java:233)
	at org.apache.pulsar.client.admin.internal.TopicsImpl.createSubscription(TopicsImpl.java:1146)
	at org.apache.pulsar.functions.worker.WorkerService.start(WorkerService.java:185)
	... 41 more
Caused by: java.lang.NullPointerException
	at org.glassfish.jersey.client.JerseyInvocation.translate(JerseyInvocation.java:697)
	at org.glassfish.jersey.client.JerseyInvocation.lambda$invoke$1(JerseyInvocation.java:632)
	at org.glassfish.jersey.client.JerseyInvocation.call(JerseyInvocation.java:654)
	at org.glassfish.jersey.client.JerseyInvocation.lambda$runInScope$3(JerseyInvocation.java:648)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:205)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:390)
	at org.glassfish.jersey.client.JerseyInvocation.runInScope(JerseyInvocation.java:648)
	at org.glassfish.jersey.client.JerseyInvocation.invoke(JerseyInvocation.java:631)
	at org.glassfish.jersey.client.JerseyInvocation$Builder.method(JerseyInvocation.java:434)
	at org.glassfish.jersey.client.JerseyInvocation$Builder.put(JerseyInvocation.java:318)
	at org.apache.pulsar.client.admin.internal.TopicsImpl.createSubscription(TopicsImpl.java:1144)
	... 42 more

Error:  shutdown(org.apache.pulsar.functions.worker.PulsarFunctionLocalRunTest)  Time elapsed: 30.134 s  <<< FAILURE!
org.apache.pulsar.broker.PulsarServerException: org.apache.pulsar.broker.PulsarServerException: java.util.concurrent.TimeoutException
	at org.apache.pulsar.broker.PulsarService.close(PulsarService.java:377)
	at org.apache.pulsar.functions.worker.PulsarFunctionLocalRunTest.shutdown(PulsarFunctionLocalRunTest.java:283)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.MethodInvocationHelper.invokeMethodConsideringTimeout(MethodInvocationHelper.java:61)
	at org.testng.internal.ConfigInvoker.invokeConfigurationMethod(ConfigInvoker.java:366)
	at org.testng.internal.ConfigInvoker.invokeConfigurations(ConfigInvoker.java:320)
	at org.testng.internal.TestInvoker.runConfigMethods(TestInvoker.java:701)
	at org.testng.internal.TestInvoker.runAfterGroupsConfigurations(TestInvoker.java:677)
	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:548)
	at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:174)
	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
	at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:822)
	at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:147)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.testng.TestRunner.privateRun(TestRunner.java:764)
	at org.testng.TestRunner.run(TestRunner.java:585)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:384)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:378)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:337)
	at org.testng.SuiteRunner.run(SuiteRunner.java:286)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:53)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:96)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1218)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1140)
	at org.testng.TestNG.runSuites(TestNG.java:1069)
	at org.testng.TestNG.run(TestNG.java:1037)
	at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:135)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.executeSingleClass(TestNGDirectoryTestSuite.java:112)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.executeLazy(TestNGDirectoryTestSuite.java:123)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.execute(TestNGDirectoryTestSuite.java:90)
	at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:146)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
Caused by: org.apache.pulsar.broker.PulsarServerException: java.util.concurrent.TimeoutException
	at org.apache.pulsar.broker.web.WebService.close(WebService.java:256)
	at org.apache.pulsar.broker.PulsarService.close(PulsarService.java:279)
	... 40 more
Caused by: java.util.concurrent.TimeoutException
	at org.eclipse.jetty.util.FutureCallback.get(FutureCallback.java:130)
	at org.eclipse.jetty.util.FutureCallback.get(FutureCallback.java:30)
	at org.eclipse.jetty.server.handler.AbstractHandlerContainer.doShutdown(AbstractHandlerContainer.java:175)
	at org.eclipse.jetty.server.Server.doStop(Server.java:447)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.stop(AbstractLifeCycle.java:93)
	at org.apache.pulsar.broker.web.WebService.close(WebService.java:252)
	... 41 more

Error:  shutdown(org.apache.pulsar.functions.worker.PulsarFunctionLocalRunTest)  Time elapsed: 30.002 s  <<< FAILURE!
org.apache.pulsar.broker.PulsarServerException: org.apache.pulsar.broker.PulsarServerException: java.util.concurrent.TimeoutException
	at org.apache.pulsar.broker.PulsarService.close(PulsarService.java:377)
	at org.apache.pulsar.functions.worker.PulsarFunctionLocalRunTest.shutdown(PulsarFunctionLocalRunTest.java:283)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.MethodInvocationHelper.invokeMethodConsideringTimeout(MethodInvocationHelper.java:61)
	at org.testng.internal.ConfigInvoker.invokeConfigurationMethod(ConfigInvoker.java:366)
	at org.testng.internal.ConfigInvoker.invokeConfigurations(ConfigInvoker.java:320)
	at org.testng.internal.TestInvoker.runConfigMethods(TestInvoker.java:701)
	at org.testng.internal.TestInvoker.runAfterGroupsConfigurations(TestInvoker.java:677)
	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:548)
	at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:174)
	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
	at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:822)
	at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:147)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.testng.TestRunner.privateRun(TestRunner.java:764)
	at org.testng.TestRunner.run(TestRunner.java:585)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:384)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:378)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:337)
	at org.testng.SuiteRunner.run(SuiteRunner.java:286)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:53)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:96)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1218)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1140)
	at org.testng.TestNG.runSuites(TestNG.java:1069)
	at org.testng.TestNG.run(TestNG.java:1037)
	at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:135)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.executeSingleClass(TestNGDirectoryTestSuite.java:112)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.executeLazy(TestNGDirectoryTestSuite.java:123)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.execute(TestNGDirectoryTestSuite.java:90)
	at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:146)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
Caused by: org.apache.pulsar.broker.PulsarServerException: java.util.concurrent.TimeoutException
	at org.apache.pulsar.broker.web.WebService.close(WebService.java:256)
	at org.apache.pulsar.broker.PulsarService.close(PulsarService.java:279)
	... 40 more
Caused by: java.util.concurrent.TimeoutException
	at org.eclipse.jetty.util.FutureCallback.get(FutureCallback.java:130)
	at org.eclipse.jetty.util.FutureCallback.get(FutureCallback.java:30)
	at org.eclipse.jetty.server.handler.AbstractHandlerContainer.doShutdown(AbstractHandlerContainer.java:175)
	at org.eclipse.jetty.server.Server.doStop(Server.java:447)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.stop(AbstractLifeCycle.java:93)
	at org.apache.pulsar.broker.web.WebService.close(WebService.java:252)
	... 41 more
</pre></code>
</detail>

**Comment by lhotari** [comment]
One reason for the flakiness of this test has been #9564 . There's a fix in #9565 for that issue. It might fix the flakiness in PulsarFunctionLocalRunTest. However there could be multiple issues behind the flakiness.

**Comment by lhotari** [comment]
I have revisited PulsarFunctionLocalRunTest [in my experimental branch](https://github.com/lhotari/pulsar/commits/lh-refactor-github-action-workflows) where I'm working on the Pulsar CI GitHub Actions workflow refactoring. I was forced to revisit it as a result of a build dependency change I'm working on where Pulsar IO .nar packaging would be switched to properly specify `<packaging>nar</packaging>` for Pulsar IO connector projects. This takes the classes off the classpath and some tests were dependent on this, such as PulsarFunctionLocalRunTest. While fixing it, I have found and fixed quite a few issues in the test which I'll contribute in proper PRs asap.
